From cd2a94252b153cb47630226789338139d0063fae Mon Sep 17 00:00:00 2001
From: Pierre-Hugues Husson <phh@phh.me>
Date: Sun, 17 Oct 2021 17:17:13 -0400
Subject: [PATCH 4/8] Don't abort when failing to get real-time priority

On some devices (like OP6), for unknown reason, trying to go to realtime
fails with EPERM.
There is no good reason to actually require real-time, so don't fail
when we don't get it
This fixes gabeldorsche on OP6
Not yet legacy bluetooth stack

Change-Id: Id25dac186628e933185bdfd640498004459b375a
---
 system/btif/src/btif_a2dp_sink.cc   |  2 +-
 system/btif/src/btif_a2dp_source.cc |  2 +-
 system/osi/src/alarm.cc             | 12 +++++++++++-
 system/stack/btu/main_thread.cc     |  2 +-
 4 files changed, 14 insertions(+), 4 deletions(-)

diff --git a/system/btif/src/btif_a2dp_sink.cc b/system/btif/src/btif_a2dp_sink.cc
index ffa783cdab..8960603a31 100644
--- a/system/btif/src/btif_a2dp_sink.cc
+++ b/system/btif/src/btif_a2dp_sink.cc
@@ -196,7 +196,7 @@ bool btif_a2dp_sink_init() {
   /* Schedule the rest of the operations */
   if (!btif_a2dp_sink_cb.worker_thread.EnableRealTimeScheduling()) {
 #if defined(__ANDROID__)
-    log::fatal("Failed to increase A2DP decoder thread priority");
+    log::error("Failed to increase A2DP decoder thread priority");
 #endif
   }
   btif_a2dp_sink_cb.worker_thread.DoInThread(
diff --git a/system/btif/src/btif_a2dp_source.cc b/system/btif/src/btif_a2dp_source.cc
index b691805fa9..9643ca2a08 100644
--- a/system/btif/src/btif_a2dp_source.cc
+++ b/system/btif/src/btif_a2dp_source.cc
@@ -372,7 +372,7 @@ static void btif_a2dp_source_startup_delayed() {
   log::info("state={}", btif_a2dp_source_cb.StateStr());
   if (!btif_a2dp_source_thread.EnableRealTimeScheduling()) {
 #if defined(__ANDROID__)
-    log::fatal("unable to enable real time scheduling");
+    log::error("unable to enable real time scheduling");
 #endif
   }
   if (!bluetooth::audio::a2dp::init(&btif_a2dp_source_thread)) {
diff --git a/system/osi/src/alarm.cc b/system/osi/src/alarm.cc
index b0206d326a..d6898fbc74 100644
--- a/system/osi/src/alarm.cc
+++ b/system/osi/src/alarm.cc
@@ -667,7 +667,17 @@ static bool timer_create_internal(const clockid_t clock_id, timer_t* timer) {
   sigevent.sigev_notify = SIGEV_THREAD;
   sigevent.sigev_notify_function = (void (*)(union sigval))timer_callback;
   sigevent.sigev_notify_attributes = &thread_attr;
-  if (timer_create(clock_id, &sigevent, timer) == -1) {
+
+  int ret = timer_create(clock_id, &sigevent, timer);
+  if (ret == -1) {
+       log::error("{} failed to create timer with RT err {}... Try again without RT", __func__, strerror(errno));
+      // Recreate timer without RT priority
+      memset(&sigevent, 0, sizeof(sigevent));
+      sigevent.sigev_notify = SIGEV_THREAD;
+      sigevent.sigev_notify_function = (void (*)(union sigval))timer_callback;
+      ret = timer_create(clock_id, &sigevent, timer);
+  }
+  if (ret == -1) {
     log::error("unable to create timer with clock {}: {}", clock_id,
                strerror(errno));
     if (clock_id == CLOCK_BOOTTIME_ALARM) {
diff --git a/system/stack/btu/main_thread.cc b/system/stack/btu/main_thread.cc
index b6ba58cc9c..08174ed92f 100644
--- a/system/stack/btu/main_thread.cc
+++ b/system/stack/btu/main_thread.cc
@@ -77,7 +77,7 @@ void main_thread_start_up() {
   }
   if (!main_thread.EnableRealTimeScheduling()) {
 #if defined(__ANDROID__)
-    log::fatal("unable to enable real time scheduling");
+    log::error("unable to enable real time scheduling");
 #else
     log::error("unable to enable real time scheduling");
 #endif
-- 
2.34.1

